src/uharfbuzz/_harfbuzz.pyx

```diff
@@ -1057,3 +1057,152 @@ cdef class DrawFuncs:
+        cdef unsigned int num_links = len(links)
+        for i in range(num_links):
+            self._hb_link[i].position = links[i][0]
+            self._hb_link[i].width = links[i][1]
+            self._hb_link[i].objidx = links[i][2]
+
+cdef hb_link_t* create_links(unsigned int link_num,
+                             hb_object_t* obj_list,
+                             unsigned int length,
+                             hb_bool_t* success):
+    success[0] = 1
+    cdef hb_link_t* p = NULL
+    if link_num == 0:
+        return NULL
+    else:
+        p = <hb_link_t*>calloc(link_num, sizeof(hb_link_t))
+        if p == NULL:
+            free_allocated(obj_list, length)
+            # for virtual_links, need to free additional real_links which
+            # is created before virtual_links
+            if obj_list[length].real_links != NULL:
+                free(obj_list[length].real_links)
+            free(obj_list)
+
+            success[0] = 0
+            return NULL
+        else:
+            return p
+
+cdef void free_allocated(hb_object_t* obj_list, unsigned int length):
+    cdef unsigned int i
+    for i in range(length):
+        if obj_list[i].real_links != NULL:
+            free(obj_list[i].real_links)
+        if obj_list[i].virtual_links != NULL:
+            free(obj_list[i].virtual_links)
+
+# The whole table is represented as a Graph
+# The input graphnodes is a flat list of subtables
+# each Node(subtable) is represented by a tuple of 2 lists: real_links and virtual_links
+#
+# real_links is a list of edges for this node (offset field in the table)
+# An real edge is represented by a tuple of (int position, int width, int
+# objidx):
+#
+# - position: means relative position of the offset field in bytes from the 
+#             beginning of the subtable's C struct. e.g:
+#   a GSUB header struct in C looks like below:
+#   uint16 majorVersion
+#   uint16 minorVersion
+#   offset16 scriptListOffset
+#   offset16 featureListOffset
+#   offset16 lookupListOffset
+#   And the position for scriptListOffset is 4 which is calculated from (16+16)/8
+# - width: size of the offset (e.g 2 for offset16 and 4 for offset32)
+# - objidx: objidx is the index of the subtable in graph/tree generated by
+#           postorder traversal
+#
+# virtual_links is a list of virtual edges, they are not real offset links,
+# a virtual_link specifies an ordering constraint that harfbuzz packing must follow.
+# A virtual edge would have 0 offsetSize, 0 position, and a real objidx.
+# e.g, if Node A has a virtual edge with objidx b(corresponding Node is B),
+# then that means Node B is always packed after Node A in the final serialized order
+def hb_repack(subtables: List[bytes],
+              graphnodes: List[Tuple[List[Tuple[int, int, int]],
+                                     List[Tuple[int, int, int]]
+                                     ]]) -> bytes:
+    if len(subtables) != len(graphnodes):
+        return bytes(b'')
```

Comment:
`b''` is already a bytes literal, you don't need to cast to `bytes(b'')`

Reply:
I don't like that errors are communicated to the caller with this implicit empty bytes string. I think you should raise python exceptions with error messages that make sense and let the caller handle them. In this case, it should be a ValueError probably because the contract of this method is that the two input lists must have the same length